### PR TITLE
feat(api): restore deleted copyrights

### LINKED ISSUE #2468: Copyright File-view
File Details (Copyright/email/url/Author, ECC, IPRA) Content, Count, finding, activated/deactivated, delete + User findings with  description, text finding, comment for both activated and deactivated. With all decision options

### PR DESCRIPTION
**Pull request by soham4abc**
## Description

Restores deleted copyright statements for a file

## How to test

send a `patch` request to the endpoint `file/upload/{uploadId}/item/{itemid}/copyright{copyrightHash}`

## Screenshots

Screenshot from 2023-06-21 11-58-18

closes #2468

<a href="https://gitpod.io/#https://github.com/fossology/fossology/pull/2486"></a>

### CODE REVIEW

File: src/www/ui/api/Controllers/CopyrightController.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  public function __construct($container)
+  {
+    parent::__construct($container);
+    $this->container = $container;
+    $this->clearingDao = $this->container->get('dao.clearing');
+    $this->licenseDao = $this->container->get('dao.license');
+    $this->uploadDao = $container->get('dao.upload');
+    $this->restHelper = $container->get('helper.restHelper');
+    $this->copyrightDao = $container->get('dao.copyright');
+    $this->CopyrightHist = $this->restHelper->getPlugin('ajax-copyright-hist');
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CopyrightController.php:72]: Please rebase to resolve changes required in initializer.

File: src/www/ui/api/Controllers/CopyrightController.php [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+   * Restore copyrights for a particular file
+   *
+   * @param  ServerRequestInterface $request
+   * @param  ResponseHelper         $response
+   * @param  array                  $args
+   * @return ResponseHelper
+   */
+
+
+  public function RestoreFileCopyrights($request, $response, $args)
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CopyrightController.php:85]: Please follow code style
```suggestion
   */
  public function restoreFileCopyrights($request, $response, $args)
```

File: src/www/ui/api/Controllers/CopyrightController.php [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  {
+    try {
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CopyrightController.php:87]: Please remove `try/catch` as discussed.

File: src/www/ui/api/Controllers/CopyrightController.php [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+      $uploadDao = $this->restHelper->getUploadDao();
+      $uploadTreeId = intval($args['itemId']);
+      $copyrightHash = ($args['hash']);
+      $userId = $this->restHelper->getUserId();
+      $UploadTreeTableName = $uploadDao->getUploadtreeTableName($uploadTreeId);
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CopyrightController.php:92]:
```suggestion
      $uploadTreeTableName = $uploadDao->getUploadtreeTableName($uploadTreeId);
```

File: src/www/ui/api/Controllers/CopyrightController.php [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+      $cpTable = $this->CopyrightHist->getTableName('statement');
+      $returnVal = null;
+      $item = $uploadDao->getItemTreeBounds($uploadTreeId, $UploadTreeTableName);
+
+      if (!$this->dbHelper->doesIdExist($uploadDao->getUploadtreeTableName($uploadTreeId), "uploadtree_pk", $uploadTreeId)) {
+        $returnVal = new Info(404, "Item does not exist", InfoType::ERROR);
+      } else if (!$this->dbHelper->doesIdExist("copyright", "hash", $copyrightHash)) {
+        $returnVal = new Info(404, "License does not exist", InfoType::ERROR);
+      }
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CopyrightController.php:101]: Please move checks before accessing variables.

File: src/www/ui/api/documentation/openapi.yaml [hunk 6]
```diff
@@ -1544,6 +1544,54 @@ paths:
                 $ref: '#/components/schemas/Info'
         default:
           $ref: '#/components/responses/defaultResponse'
+  /file/upload{id}/item/{itemId}/copyright/{hash}:
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:1547]: Please update the path to be consistent with other endpoints.

File: src/www/ui/api/documentation/openapi.yaml [hunk 7]
```diff
@@ -1544,6 +1544,54 @@ paths:
...
+      summary: Restores a copyright for a file
+      description: Restores a copyright statement for a particular file
+      responses:
+        '200':
+          description: OK
+          headers:
+            X-Total-Pages:
+              description: Total number of pages which can be generated based on limit
+              schema:
+                type: integer
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:1580]: The header is irrelevant here.

File: src/www/ui/api/documentation/openapi.yaml [hunk 8]
```diff
@@ -2640,6 +2688,21 @@ components:
...
+          description: Status code
+          example: 200
+        content:
+          type: string
+          description: Success message
+          example: "Successfully restored Copyright."
+        type:
+          type: string
+          description: type of message
+          example: "INFO"
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:2705]: Response is same as `Info`.

**Review by GMishx — CHANGES_REQUESTED** [review]

File: src/www/ui/api/documentation/openapi.yaml [hunk 9]
```diff
@@ -1960,7 +1983,7 @@ paths:
         content:
           application/json:
             schema:
-              $ref: '#/components/schemas/SetCopyrightInfo'
+             $ref: '#/components/schemas/SetCopyrightInfo'
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:1986]:
```suggestion
              $ref: '#/components/schemas/SetCopyrightInfo'
```

**Review by GMishx — CHANGES_REQUESTED** [review]

**Comment by soham4abc** [comment]
Re-based and updated lint change.

Please do have a look @GMishx

File: src/www/ui/api/documentation/openapi.yaml (hunk 9 above)
Comment (GMishx) [reply src/www/ui/api/documentation/openapi.yaml]: Uh, seems like you removed one space instead of adding (as suggested).

**Review by GMishx — CHANGES_REQUESTED** [review]

**Comment by GMishx** [comment]
@soham4abc please rebase the branch for merging.

**Comment by soham4abc** [comment]
I have re-based the changes!!

Please do have a look @GMishx

**Review by GMishx — APPROVED** [review]
Changes looks good.

**Review by GMishx — APPROVED** [review]
Tested, working as expected.

**Pull request merged**